README.md
<p align="center">
  <img alt="logo" src="/readme-img/logo.png?raw=true">
</p>
<p align="center">
  <a href="https://github.com/etacarinaea/startpage/releases"><img src="https://img.shields.io/github/package-json/v/etacarinaea/startpage.svg?color=%23C9A0DC&label=Version" /></a>
  <a href="https://addons.mozilla.org/en-US/firefox/addon/square-startpage/"><img src="https://img.shields.io/badge/Firefox-Download-%23FF9400.svg" /></a>
  <a href="https://chrome.google.com/webstore/detail/startpage/odmjpppdiaenedpilgldapgbodindjak"><img src="https://img.shields.io/badge/Chrome%2FChromium-Download-%234285F4.svg" /></a>
</p>
<br>

Quickly access your favourite websites whenever you open a new tab


## Key Features

* Easy customisation using a built-in configuration menu, or alternatively a
  JSON file
* Keyboard controls
* Customisable search function
* Export/Import function to backup your config or to use the same one on
  multiple devices


<br><br>
<p align="center">
  <img alt="example screenshot" src="/readme-img/screenshot-20190601.png?raw=true" width="850px">
</p>
<br><br>


## Installation

Install the [Firefox add-on][1] or the [Chrome add-on][2].

If you're using a different browser:<br>
Download a [release][3], download the [zip][4] of the latest version on GitHub,
or just clone the repository. Your browser has to support the
[WebExtensions API][5] or you'll have to download an older version, find a way
to overwrite the newtab page of your browser and select `index.html` as the new
one.

##### GitHub Pages Preview

The startpage is also hosted on
[GitHub Pages](http://etacarinaea.github.io/startpage/) for demonstration
purposes. You can set your newtab URL to that site if you like, but keep in mind
that the version hosted there is an older one from the one currently available
as a browser extension.


## Usage

The config menu can be opened any time by clicking the gear icon in the
bottom-right corner of the screen, or by writing `-config` into the search
square.

The arrow keys can be used to navigate the squares: left and right will move to
a different square and up and down will move between links in a square. The
selected link can then be visited by pressing enter.

The tab key automatically selects the search field.

To use the keyboard controls the newtab page has to have focus. This
unfortunately isn't the case when opening a new tab by default in most browsers,
and I'm not aware of any workarounds, other than clicking inside the window or
pressing tab until the page gets focus.


## Configuration

The startpage can be configured by using a configuration file or by using the
built-in menu.<br>


## Contact

If you find a bug or run into problems please create an issue [here][6].


## More examples

<p align="center">
  <img alt="example screenshot" src="/readme-img/screenshot-29032017-1.png?raw=true" width="400px">
  <img alt="example screenshot" src="/readme-img/screenshot-20191004.png?raw=true" width="400px"><br>
  <img alt="example screenshot" src="/readme-img/screenshot-29032017-2.png?raw=true" width="400px">
  <img alt="example screenshot" src="/readme-img/screenshot-29032017-4.png?raw=true" width="400px">
</p>


[1]: https://addons.mozilla.org/en-US/firefox/addon/square-startpage/
[2]: https://chrome.google.com/webstore/detail/startpage/odmjpppdiaenedpilgldapgbodindjak
[3]: https://github.com/etacarinaea/startpage/releases
[4]: https://github.com/etacarinaea/startpage/archive/master.zip
[5]: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions
[6]: https://github.com/fuyuneko/startpage/issues


CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/),
and this project adheres to [Semantic Versioning](https://semver.org/).

## [1.12.0] - 2022-2-2
### Fixed
* Enter not working when search square is missing
### Added
* `link spacing` option

## [1.11.4] - 2021-5-10
### Fixed
* Fixed arrow keys not working

## [1.11.3] - 2021-4-10
### Fixed
* Fixes an error when `searchsquare` is undefined and `alwaysopen == true`

## [1.11.2] - 2020-8-8
### Changed
* Made duckduckgo the default search engine in the default config

## [1.11.1] - 2020-6-20
### Fixed
* Fixed the help popup for search options always showing in chrome

## [1.11.0] - 2020-5-15
### Added
* This changelog file
* npm to make "building" easier
  * `patch/` contains diffs for integrating the
    [webextension-polyfill](https://github.com/mozilla/webextension-polyfill)
  * `npm run build-ff` creates a zip containing the startpage (for firefox)
  * `npm run build-ch` creates a zip containing the patched startpage with the
    webextension-polyfill (for chrome)
* Version string in config.json; the current version of the config file
* Automatically update config based on config version string
### Changed
* Configuration data is now saved in `storage.local` instead of
  `localStorage`, as is conventional for browser extensions

## [1.10.0] - 2020-2-22
### Added
* Gear button in the bottom-right corner that opens the config menu when
  clicked
* `hide_gear_button` config option; when true hides the gear icon until hovered
  over
### Removed
* `use_json_file` config option

## [1.9.2] - 2020-2-5
### Changed
* Disable background image in default config

## [1.9.1] - 2020-2-3
### Fixed
* Fix `alwaysopen` config option not working

## [1.9.0] - 2020-1-17
### Changed
* Reworked config options for background images
  * `images`: An URL to an image (local files don't work currently)
  * `size`: The image size, e.g. `500px 250px`, `cover`
  * `filter`: An optional filter, e.g. `opacity(0.2)`


[1.12.0]: https://github.com/etacarinaea/startpage/compare/v1.11.4...v1.12.0
[1.11.4]: https://github.com/etacarinaea/startpage/compare/v1.11.3...v1.11.4
[1.11.3]: https://github.com/etacarinaea/startpage/compare/v1.11.2...v1.11.3
[1.11.2]: https://github.com/etacarinaea/startpage/compare/v1.11.1...v1.11.2
[1.11.1]: https://github.com/etacarinaea/startpage/compare/v1.11.0...v1.11.1
[1.11.0]: https://github.com/etacarinaea/startpage/compare/v1.10.0...v1.11.0
[1.10.0]: https://github.com/etacarinaea/startpage/compare/v1.9.2...v1.10.0
[1.9.2]: https://github.com/etacarinaea/startpage/compare/v1.9.1...v1.9.2
[1.9.1]: https://github.com/etacarinaea/startpage/compare/v1.9.0...v1.9.1
[1.9.0]: https://github.com/etacarinaea/startpage/compare/v1.8.2...v1.9.0


js/configmenu.js
function ConfigObject(items){
    for(var i in items){
        this[i] = {name: items[i]};
    }
}

var boolItems = {
    borders: "Borders",
    alwaysopen: "keep all squares open",
    allow_version_check: "allow checking for new versions",
    use_json_file: "use config.json instead of this menu"
}
var bool = new ConfigObject(boolItems);

var styleItems = {
    heading_font: "Heading Font",
    link_font: "Link Font",
    heading_font_size: "Heading Font Size",
    link_font_size: "Link Font Size",
    background: "Background",
    foreground: "Foreground",
    heading_color: "Heading Color",
    link_color: "Link Color",
    border_color: "Border Color",
    border_width_normal: "Border Width",
    border_width_hovered: "Border Width (Hovered)",
    search_color: "Search Color",
    search_bg_color: "Search Background Color"
};
var style = new ConfigObject(styleItems);

var extItems = {
    images: "Images",
    bottom: "Bottom",
    right: "Right",
    height: "Height",
    width: "Width",
    opacity: "Opacity"
};
var ext = new ConfigObject(extItems);



function configmenuInit(callback){
    $.getJSON("config.json", function(data){
        if(data.bool.privateMode == true){
            loadConfig(data, callback);
        }else if(localStorage.use_json_file == "true" || localStorage.config == undefined){
            pipe(data, callback);
        }else{
            pipe(JSON.parse(localStorage.config), callback);
        }
    });
}

// separate function so it wont execute before jQuery.getJSON has finished
function pipe(data, callback){
    // create initial menu, config menu or load config on window load
    if(localStorage.config == undefined){
        initmenu = new Menu("Init-Menu", 1, 550, 350);
        var initbuttons = initmenu.split(
                ["Use files.",
                 "Use configuration menu."],
                ["Use the config.json file located in the startpage's root directory.",
                 "Use a GUI to easily configure the startpage's style. Has import/export function."]);
        initbuttons[0].addEventListener("click", function(){
            loadConfig(data, callback);
            initmenu.kill();
        });
        initbuttons[1].addEventListener("click", function(){
            createMenu(data, callback);
            initmenu.kill();
        });
    }else if(callback == undefined){
        createMenu(data, callback);
    }else{
        loadConfig(data, callback);
    }
}

function createMenu(data, callback){
    // create menu and categories
    configmenu = new Menu("Config-Menu", 0, 110, 110);
    var boolcategory = configmenu.appendCategory("bool", 0, undefined);
    var stylecategory = configmenu.appendCategory("style", 0, "General");
    var extcategory = configmenu.appendCategory("ext", 1, "Mascot");

    if(!data){
        var data = {bool:"",style:"",ext:""};
    }

    // append options to categories
    for(var key in bool){
        configmenu.categories[0].appendOption(bool[key].name, key, 0, callback, data.bool[key]);
    }
    for(var key in style){
        configmenu.categories[1].appendOption(style[key].name, key, 1, callback, data.style[key]);
    }
    for(var key in ext){
        configmenu.categories[2].appendOption(ext[key].name, key, 1, callback, data.ext[key]);
    }

    var saveButton = configmenu.appendButton("save", "#99bb99");
    saveButton.addEventListener("click", function(){
        saveConfig(callback);
        configmenu.kill();
    });
    var exportButton = configmenu.appendButton("export", "#9999bb");
    exportButton.addEventListener("click", function(){
        exportConfig();
    });
    var importButton = configmenu.appendButton("import", "#bb9999");
    importButton.addEventListener("click", function(){
        importConfig(callback);
    });
}


function importConfig(callback){
    var importinput = document.createElement("input");
    importinput.setAttribute("type", "file");
    importinput.setAttribute("name", "importinput");

    configmenu.heading.appendChild(importinput);

    importinput.addEventListener("change", function(e){
        var file = importinput.files[0];

        var reader = new FileReader();
        reader.readAsText(file);

        reader.onload = function(e){
            loadConfig(JSON.parse(reader.result), callback);
            configmenu.kill();
        }
    });

    importinput.click();
}

function exportConfig(){
    saveConfig();
    window.open("data:application/octet-stream;base64," + btoa(localStorage.config));
}

function saveConfig(callback){
    json = {bool:{}, style:{}, ext:{}};
    for(var key in bool){
        var elem = document.querySelector("input[name='" + key + "'");
        bool[key].value = elem.checked;
        json.bool[key] = elem.checked;
    }
    localStorage.use_json_file = document.querySelector("input[name='use_json_file'").checked;
    for(var key in style){
        var elem = document.querySelector(
                "#style input[name='" + key + "'");
        style[key].value = elem.value;
        json.style[key] = String(elem.value);
    }
    for(var key in ext){
        var elem = document.querySelector(
                "#ext input[name='" + key + "'");
        ext[key].value = elem.value;
        json.ext[key] = String(elem.value);
    }

    loadConfig(json, callback);
}


function loadConfig(data, callback){
    localStorage.config = JSON.stringify(data, undefined, 4);
    cfg = [
        data.style.heading_font,
        data.style.link_font,
        data.style.heading_font_size,
        data.style.link_font_size,
        data.style.background,
        data.style.foreground,
        data.style.heading_color,
        data.style.link_color,
        data.style.border_color,
        data.style.border_width_normal,
        data.style.border_width_hovered,
        data.style.search_color,
        data.style.search_bg_color,
        data.ext.images,
        data.ext.bottom,
        data.ext.right,
        data.ext.height,
        data.ext.width,
        data.ext.opacity
    ];
    cfg_bool = [
        data.bool.borders,
        data.bool.alwaysopen,
        data.bool.mascot,
        data.bool.allow_version_check
    ];
    localStorage.cfg = cfg, localStorage.cfg_bool = cfg_bool;
    var span = $("span");
    var a = $("a");
    var popup = $("#popup");
    var sqr = $(".sqr");
    span.css("fontFamily", cfg[0]);
    a.css("fontFamily", cfg[1]);
    popup.css("fontFamily", cfg[1]);
    span.css("fontSize", cfg[2]);
    a.css("fontSize", cfg[3]);
    popup.css("fontSize", cfg[3]);
    $("body").css("backgroundColor", cfg[4]);
    sqr.css("backgroundColor", cfg[5]);
    popup.css("backgroundColor", cfg[5]);
    span.css("color", cfg[6]);
    a.css("color", cfg[7]);
    popup.css("color", cfg[7]);
    sqr.css("borderColor", cfg[8]);
    sqr.css("borderWidth", cfg[9]);
    popup.css("borderTop", cfg[9] + " solid " + cfg[8]);
    var searchinput = $("#searchinput");
    if(searchinput.length){
        searchinput.css("color", cfg[11]);
        searchinput.css("backgroundColor", cfg[12]);
    }
    var bgimg = $("#bgimg");
    if(cfg_bool[2]){
        bgimg.css("backgroundImage", "url('" +
                cfg[13][Math.floor(Math.random()*cfg[13].length)] + "')");
        bgimg.css("bottom", cfg[14]);
        bgimg.css("right", cfg[15]);
        bgimg.css("height", cfg[16]);
        bgimg.css("width", cfg[17]);
        bgimg.css("opacity", cfg[18]);
    }else{
        bgimg.css("backgroundImage", "");
    }

    if(callback){
        callback();
    }
}

